src/Factory/ServerRequestFactory.php

```diff
@@ -4,35 +4,178 @@
+    public function createServerRequestFromArray(array $server): ServerRequestInterface
     {
         if (!isset($server['REQUEST_METHOD'])) {
-            throw new \InvalidArgumentException('Cannot determine HTTP method');
+            throw new InvalidArgumentException('Cannot determine HTTP method');
         }
-        // TODO: find a MUCH better way
         $method = $server['REQUEST_METHOD'];
-        $SERVER = $_SERVER;
-        $_SERVER = $server;
-        // Until https://github.com/guzzle/psr7/pull/116 is resolved
-        if (!isset($_SERVER['HTTPS'])) {
-            $_SERVER['HTTPS'] = 'off';
+
+        $uri = (new UriFactory())->createUriFromArray($server);
+        if ($uri->getScheme() === '') {
+            $uri = $uri->withScheme('http');
+        }
+
+        $protocol = isset($server['SERVER_PROTOCOL']) ? str_replace('HTTP/', '', $server['SERVER_PROTOCOL']) : '1.1';
+
+        return new ServerRequest($method, $uri, [], null, $protocol, $server);
+    }
+
+    /**
+     * Create a new server request from a set of arrays.
+     *
+     * @param array $server Typically $_SERVER or similar structure.
+     * @param array $cookie Typically $_COOKIE or similar structure.
+     * @param array $get    Typically $_GET or similar structure.
+     * @param array $post   Typically $_POST or similar structure.
+     * @param array $files  Typically $_FILES or similar structure.
+     *
+     * @throws InvalidArgumentException If no valid method or URI can be determined.
+     *
+     * @return ServerRequestInterface
+     */
+    public function createServerRequestFromArrays(
+        array $server,
+        array $cookie,
+        array $get,
+        array $post,
+        array $files
+    ): ServerRequestInterface {
+        if (!isset($server['REQUEST_METHOD'])) {
+            throw new InvalidArgumentException('Cannot determine HTTP method');
+        }
+        $method = $server['REQUEST_METHOD'];
+
+        $uri = (new UriFactory())->createUriFromArray($server);
+        if ($uri->getScheme() === '') {
+            $uri = $uri->withScheme('http');
+        }
+
+        $protocol = isset($server['SERVER_PROTOCOL']) ? str_replace('HTTP/', '', $server['SERVER_PROTOCOL']) : '1.1';
+
+        $headers = function_exists('getallheaders') ? getallheaders() : [];
+
+        $serverRequest = new ServerRequest($method, $uri, $headers, null, $protocol, $server);
+
+        return $serverRequest
+            ->withCookieParams($cookie)
+            ->withQueryParams($get)
+            ->withParsedBody($post)
+            ->withUploadedFiles(self::normalizeFiles($files));
+    }
+
+    /**
+     * Create a new server request from the current environment variables.
+     *
+     * @return ServerRequestInterface
+     */
+    public function createServerRequestFromGlobals(): ServerRequestInterface
+    {
+        $server = $_SERVER;
+        if (false === isset($server['REQUEST_METHOD'])) {
+            $server['REQUEST_METHOD'] = 'GET';
```

Comment:
For simplicity of `(new ServerRequestFactory())->createServerRequestFromGlobals()`, we make sure there is a method available. If users want to use `createServerRequestFromArrays` directly, they will have to make sure of it themselves.